### PR TITLE
[SYCL][CUDA] Don't infer memcpy src/dst pitches from copied region

### PR DESCRIPTION
**Pull request by psalz**
The SYCL CTS recently merged an [extensive set of additional tests](https://github.com/KhronosGroup/SYCL-CTS/blob/dc506f54f6905ada8b141c7c056b9775775f5e48/tests/handler/handler_copy.cpp) for the explicit copy operations APIs. Running these tests with the DPC++ CUDA backend reveals that copies are failing for strided accessors (i.e. copying from/to an accessor that has an offset and/or a range smaller than the buffer).

Previously, when passing 0 for row/slice pitches for source and/or destination to `commonEnqueueMemBufferCopyRect`, the function would infer the row/slice pitches from the copied region. This does not work when copying with an offset, as `cuMemcpy3DAsync` checks whether the pitches are at least as large as the offset + copied range, otherwise returning CUDA_ERROR_INVALID_VALUE.

The [CUDA documentation](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g79f4f3fde6ae0f529568d881d9e11987) says:

> * If specified, srcPitch must be greater than or equal to WidthInBytes + srcXInBytes, and dstPitch must be greater than or equal to WidthInBytes + dstXInBytes.
> * If specified, srcHeight must be greater than or equal to Height + srcY, and dstHeight must be greater than or equal to Height + dstY.

For 1D / 2D copies, no / not all pitches are required, and can instead be specified as 0. For the `srcHeight` / `dstHeight` parameters the documentation mentions this explicitly, but from the wording above ("If specified"), I'm assuming the same applies for `srcPitch` / `dstPitch`.

I've thus removed the fallback mechanism and replaced it with assertions to check that for 2D and 3D copies, non-zero pitches are provided. I'm not sure why this mechanism exists in the first place, however it is also [present](https://github.com/intel/llvm/blob/c8544f1c1c22529c0a8d16e203b395b4d2ef55e0/sycl/plugins/level_zero/pi_level_zero.cpp#L4580) in the Level Zero implementation. In my testing, I have not encountered any zero pitches being passed, except when they are not required (i.e. for 1D/2D copies).

### CODE REVIEW

**Comment by bader** [comment]
+@steffenlarsen

**Comment by steffenlarsen** [comment]
Nice catch and I agree with your assessment. Have you tested with both the CTS and [intel/llvm-test-suite](https://github.com/intel/llvm-test-suite)? It's some time ago, but I believe there was a reason for adding these cases, but maybe the uses have changed since.

**Review by smaslov-intel — DISMISSED** [review]
LGTM

**Comment by bader** [comment]
@psalz, there are 6 new [llvm-test-suite](https://github.com/intel/llvm-test-suite/tree/intel/SYCL) tests failing with this change:
SYCL :: Basic/access_to_subset.cpp
SYCL :: Basic/accessor/accessor.cpp
SYCL :: Basic/buffer/buffer.cpp
SYCL :: Basic/buffer/reinterpret.cpp
SYCL :: Plugin/enqueue-arg-order-buffer.cpp
SYCL :: Tracing/buffer_printers.cpp

Could you take a look, please?
I think they might be updated to align with these changes.

**Comment by psalz** [comment]
Sorry, I must've overlooked the note on end-to-end tests in the contributing guide (I was not aware of them). I did run them now and I think I have a better understanding of what is happening.

While I did not examine all tests in detail, the first one ([BASIC/access_to_subset.cpp](https://github.com/intel/llvm-test-suite/blob/e1607bfeb08f2936c0ffa3442c4d8dca095aa5a9/SYCL/Basic/access_to_subset.cpp)) already revealed an important difference in the way certain copies are handled (and I suspect all other tests fail because of a variation on that). In particular, that test uses an automatic writeback on buffer destruction to get device results back to the host. Internally, this means a copy from the 2D device buffer to the given (1D) host pointer has to be done.

While I did not address this specifically in my original post, the CTS actually also tests all kinds of combinations of dimensionalities for the source and destination of explicit copies (e.g., copying from a 2D buffer into a 3D buffer). As the tests for explicit copies in the CTS now all pass, I compared what is happening between an explicit 2D -> 1D copy there, versus the implicit buffer writeback. As it turns out, both first go through the same path within [`copyD2H`](https://github.com/intel/llvm/blob/ff942e9c92d211e759ec55a95dc118d6d4b647ee/sycl/source/detail/memory_manager.cpp#L376). However, there is a difference in how the host pointer is handled: While the explicit copy treats the destination pointer as a 2D buffer with appropriate row pitch, for a writeback it is instead handled as a 1D destination with a 2D source and copy region, which implies a `dst_row_pitch` of zero and then in turn causes the assertion to fail.

**TLDR:** There seems to be some inconsistencies within the DPC++ runtime as to how buffer-to-pointer copies are handled. I'm afraid that I do not possess sufficient knowledge of the codebase to address those, however. So the more conservative and pragmatic route would be to simply reinstate the fallback for zero pitches, but to also incorporate the Y-offset for source and destination into the slice pitch, so that CUDA doesn't complain. With this, both the CTS and E2E tests seem to pass!

**Comment by bader** [comment]
@romanovvlad, I recall there were some changes related to similar copies. Can you suggest who can help with addressing these issue? Please, read [this comment](https://github.com/intel/llvm/pull/3563#issuecomment-823362443) for more details.

**Comment by bader** [comment]
@romanovvlad, @intel/llvm-reviewers-runtime, could you help here, please?

**Comment by psalz** [comment]
I've added a commit that reverts the previous changes, instead opting to also consider src/dst offsets for the inferred pitches. As previously mentioned, this is a more conservative change that doesn't require larger adjustments throughout the runtime, while still passing both the E2E tests as well as the CTS.

If you are fine with this change, I can squash it into a single commit and rebase onto current head.

**Comment by bader** [comment]
> I can squash it into a single commit and rebase onto current head.

Usually it's done by maintainers during pull request merge. It's better to avoid rebasing patches in PR to keep track of the comments.

**Review by bader — APPROVED** [review]

**Comment by psalz** [comment]
> Usually it's done by maintainers during pull request merge. It's better to avoid rebasing patches in PR to keep track of the comments.

Okay sure (although I've noticed that GitHub seems to be smart enough to retain comments at their original positions for most rebases and forced pushes). I would still recommend to squash though, as the first commit is just noise at this point.

**Comment by bader** [comment]
@intel/llvm-reviewers-cuda, ping.
